pkg/driver/controller.go

```diff
@@ -63,6 +64,15 @@ type controller struct {
 
 	k8sNodeInformer cache.SharedIndexInformer
 	zfsNodeInformer cache.SharedIndexInformer
+
+	volMutexes sync.Map
+}
+
+func (cs *controller) LockVolume(volume string) func() {
+	value, _ := cs.volMutexes.LoadOrStore(volume, &sync.Mutex{})
```

Comment:
Just wondering, are entries automatically GC'd?
Otherwise this may grow "forever" ?

Reply:
No, they are not getting GC'd. The Mutex can only be removed from the Map if there are no other references to the Mutex. Im not sure if this relatively small leak (the volume string key is probably larger than the mutex) is worth the additional complexity.

Reply:
Well, it might grow to large extents? Also what about deleleted volume, the entries for them will also stay in the map right?

Reply:
Seems most drivers on kubernetes-csi actually use the same exact implementation: https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/pkg/common/volume_lock.go
Which handles it by not using `N` locks but rather keep a list of volumes and deleting the volume entry.

Reply:
Yeah guess we'd need to do the same here as well. Perhaps that's also a better approach, letting the caller retry.
Should we use this new way instead? @Lucaber @Abhinandan-Purkait

Reply:
I have looked at a few different Implementations. TopoLVM also uses a single map to add and remove keys but using a sync.Cond that locks the second concurrent call like my implementation: https://github.com/topolvm/topolvm/blob/main/internal/driver/lock.go This seems to be the best way to handle this.

Alternatively we could also use the keymutex implementation by kubernetes that used a few mutexes which are being shared between volumes. But this would require special handling to ensure that the snapshot and parent volume are using different mutexes (to prevent deadlocking). https://github.com/kubernetes/utils/blob/master/keymutex/hashed.go

Reply:
The topolvm one looks reasonable, wdyt?